### DISCUSSION TITLE
CICD support for Bitbucket Repository

### DISCUSSION THREAD

**Discussion by HarshalSanap007** [GoogleCloudPlatform/cloud-foundation-fabric#2763] [General]
Hey @ludoo , 
Do we have the CICD implementation support for Bitbucket repo and Cloudbuild as CICD tool?

**Comment by ludoo** [comment]
There was some work last year to add Bitbucket support but it stalled. Cloud build we used to have, we dropedd support once GCP Source Repos were deprecated. We might reconsider both if there's actual need, of course.

**Reply by HarshalSanap007** [reply]
Hey @ludoo , thanks for confirming ,I had one doubt which this bootstrap stage should be done manually or it can be done via CICD as well ?

**Reply by ludoo** [reply]
bootstrap needs to be run manually at least the first two times: first as a user, then as the service account with providers/backend file in place

after that you can plug it into a repo/pipeline

**Reply by HarshalSanap007** [reply]
Hey @ludoo , thanks for your response, every time I test it I get new errors every time mostly related to Iam. so this one error is really troubling me , I've added provider.tf in which I'm passing service account in following manner

terraform {
  backend "gcs" {
    bucket                      = "fast-fabric-tf-state"
    prefix = "terraform/state/dev-org"
    impersonate_service_account = "sa-terraform-user@bootstrap-prj-ff.iam.gserviceaccount.com"
  }
}
provider "google" {
  impersonate_service_account = "sa-terraform-user@bootstrap-prj-ff.iam.gserviceaccount.com"
}
provider "google-beta" {
  impersonate_service_account = "sa-terraform-user@bootstrap-prj-ff.iam.gserviceaccount.com"
}

It throws the error that 403: some permissions are required and when I pass those through console it and then run plan again then it shows that service account permission which I granted is shown under ( -  means will be removed )

**Reply by ludoo** [reply]
you should use the providers file and service account we generate...

**Reply by HarshalSanap007** [reply]
> bootstrap needs to be run manually at least the first two times: first as a user, then as the service account with providers/backend file in place
> 
> after that you can plug it into a repo/pipeline

Does that means it should not be deployed as CICD from start, once all resource are created further customizations can be part of CICD , is that what you mean here?

Thanks?

**Reply by HarshalSanap007** [reply]
> you should use the providers file and service account we generate...

This I'm trying to add for bootstrap process .

**Reply by ludoo** [reply]
Sorry I can't follow what you're writing. Just make sure you go through the flow describe in the bootstrap stage documentation, without devising your own method to deploy it which won't work and we don't support.

**Reply by HarshalSanap007** [reply]
Extremly sorry that you got confused here

**Comment by HarshalSanap007** [comment]
I Might have confused you - I will try to explain again here-
So I want to create landing zone with fast fabric modules, I've tried each step manually like going to each stage and doing terraform init and apply I've applied  and created resources for every stage .
Now Im trying to achieve this same with CICD deployment creating individual repo for each stage so - 
my questions - 
1: Do we need to run 0-bootstrap stage manually or it can be directly run under CICD?
2: If we can run under CICD , as we are migrating the state from local location "fast-config" to iac bucket, instead of local outputs location can we directly pass remote location "outputs_location = "~/fast-config-ofw"?

**Reply by ludoo** [reply]
You need to have a working bootstrap already set up, and configure the `cicd` variable. Once you do that we create additional service accounts, and the workflow files for Github or Gitlab.

State is always in the same GCS bucket. Fast config is used for output/provider files, but the same files exist in the GCS outputs bucket and are fetched from there by the workflows. You should not use `outputs_location` once you switch to CI/CD.

Again, please read through our docs. :)

**Reply by HarshalSanap007** [reply]
Thanks it helped
